### PR TITLE
MAINT: extend delete single value optimization

### PR DESCRIPTION
**Pull request by DevinShanahan**
Allow arrays of shape (1,) for `np.delete`'s `obj` parameter to utilize the optimization for a single value. See #16685.

```python
In [2]: arr = np.arange(3000000).reshape(1000, 3000)      
                      
In [3]: %timeit np.delete(arr, obj=0, axis=1)                                   
6.91 ms ± 352 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: %timeit np.delete(arr, obj=[0], axis=1)                                 
6.72 ms ± 245 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [5]: zero = np.array([0])                                                    

In [6]: %timeit np.delete(arr, obj=zero, axis=1)                                
6.77 ms ± 242 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

### CODE REVIEW

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -4368,6 +4368,19 @@ def delete(arr, obj, axis=None):
...
     if isinstance(obj, (int, integer)) and not isinstance(obj, bool):
+        single_value = True
+    else:
+        single_value = False
+        _obj = obj
+        obj = np.asarray(obj)
+        if obj.size == 0 and not isinstance(_obj, np.ndarray):
+            obj = obj.astype(intp)
+
+        if obj.shape == (1,):
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:4379]:
```suggestion
        elif obj.shape == (1,):
```

**Review by mattip — DISMISSED** [review]
Closes #16685

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -4368,6 +4368,19 @@ def delete(arr, obj, axis=None):
...
+        single_value = True
+    else:
+        single_value = False
+        _obj = obj
+        obj = np.asarray(obj)
+        if obj.size == 0 and not isinstance(_obj, np.ndarray):
+            obj = obj.astype(intp)
+
+        elif obj.shape == (1,):
+            obj = obj.item()
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:4380]: `item()` makes me worry here...

Comment (eric-wieser) [reply numpy/lib/function_base.py:4380]: Does this patch change the behavior of ` np.delete(np.arange(10), np.array([1], dtype=object))`?

File: numpy/lib/function_base.py [hunk 3]
```diff
@@ -4368,6 +4368,19 @@ def delete(arr, obj, axis=None):
...
     if isinstance(obj, (int, integer)) and not isinstance(obj, bool):
+        single_value = True
+    else:
+        single_value = False
+        _obj = obj
+        obj = np.asarray(obj)
+        if obj.size == 0 and not isinstance(_obj, np.ndarray):
+            obj = obj.astype(intp)
+
+        elif obj.shape == (1,):
```
Comment (eric-wieser) [inline numpy/lib/function_base.py:4379]: You can hit a lot more paths here with 
```suggestion
        elif obj.size== 1:
```

File: numpy/lib/function_base.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/lib/function_base.py:4380]: I think `obj = obj.reshape(())` would avoid that shortcoming - please add a test!

**Comment by InessaPawson** [comment]
@DevinShanahan Do you think you could finish this PR? It would be nice to fix #16685.

**Comment by DevinShanahan** [comment]
@InessaPawson Yes, I will try to complete this as soon as possible.

File: numpy/lib/function_base.py (hunk 2 above)
Comment (DevinShanahan) [reply numpy/lib/function_base.py:4380]: Could your clarify the behavior you are concerned about for  `np.delete(np.arange(10), np.array([1], dtype=object))`. I will change it to `obj = obj.reshape(())` either way but the behavior appears the same to me. What would you like to see in the test?

Also, is there anything else you would like to see tested?

**Comment by DevinShanahan** [comment]
@InessaPawson I've re-requested a review, anything else you need from me on this at this time?

**Comment by InessaPawson** [comment]
@DevinShanahan Thank you for updating the PR! We will review it shortly.

**Comment by mattip** [comment]
Thanks @DevinShanahan

**Comment by InessaPawson** [comment]
Hi-five on merging your first pull request to NumPy, @DevinShanahan! We hope you stick around! Your choices aren’t limited to programming – you can review pull requests, help us stay on top of new and old  issues, develop educational material, work on our website, add or improve graphic design, create marketing materials, translate website content, write grant proposals, and help with other fundraising initiatives. For more info, check out: https://numpy.org/contribute
Also, consider joining our mailing list. This is a great way to connect with other cool people in our community and be part of important conversations that affect the development of NumPy: https://mail.python.org/mailman/listinfo/numpy-discussion

**Comment by seberg** [comment]
This PR broke bolean array inputs, see gh-21840 (at least I am practically certain).  Unfortunately, we need to either *carefully* check all the paths the "integer array" paths (which seem to accept booleans arrays now) or simply revert it and give it another shot some other time.

**Pull request merged**